### PR TITLE
Feature/mean token accuracy trainer #40303

### LINKED ISSUE #40303: Add mean_token_accuracy from TRL's `SFTTrainer` to `Trainer`
### Feature request

TRL's `SFTTrainer` has a metric called `mean_token_accuracy`: https://github.com/huggingface/trl/blob/main/trl/trainer/sft_trainer.py#L1095-L1123:

```

                shift_logits = outputs.logits[..., :-1, :].contiguous()
                shift_labels = inputs["labels"][..., 1:].contiguous()

                # When using Prompt Tuning, skip the virtual tokens in logits before accuracy computation, since they do
                # not correspond to actual input labels.
                shift_logits = shift_logits[:, self.num_virtual_tokens :, :]

                # Get predictions
                predictions = shift_logits.argmax(dim=-1)

                # Create mask for non-padding tokens (assuming ignore_index is -100)
                mask = shift_labels != -100

                # Calculate accuracy only on non-padding tokens
                correct_predictions = (predictions == shift_labels) & mask
                total_tokens = mask.sum()
                correct_tokens = correct_predictions.sum()

                # Gather the correct_tokens and total_tokens across all processes
                correct_tokens = self.accelerator.gather_for_metrics(correct_tokens)
                total_tokens = self.accelerator.gather_for_metrics(total_tokens)

                # Compute the mean token accuracy and log it
                total_sum = total_tokens.sum()
                accuracy = (correct_tokens.sum() / total_sum).item() if total_sum > 0 else 0.0
                self._metrics[mode]["mean_token_accuracy"].append(accuracy)
```

I like this metric and I'd like for this metric to be optionally be added to `transformer`'s `Trainer`.

### Motivation

(Mean) Token Accuracy is a useful additional metric that I'd like the option of additionally logging.

### PR DESCRIPTION
**Pull request by badwriter123**
# What does this PR do?

Feature : Add mean token accuracy to Trainer

Fixes #40303 

### Description
1. For the file -->  src/transformers/training_args.py  Added a new boolean argument to allow users to toggle the mean token accuracy computation.

2. For the file --> src/transformers/trainer.py. Added the new method  (_compute_mean_token_accuracy) . Improvised for both Language Models and Seq2seq Models.

The feature is now ready to be used by setting compute_mean_token_accuracy=True in TrainingArguments.

### CODE REVIEW

**Comment by RylanSchaeffer** [comment]
Why are changes to gemma3n in this PR?

**Comment by RylanSchaeffer** [comment]
Similarly, why have dependencies changed?

**Comment by badwriter123** [comment]
Actually the changes other than the mean token accuracy feature were from a different pr, that has not been reviewed or merged. It got mixed up with the mean token accuracy feature.

They can be overwritten, and just the mean token accuracy feature can be reviewed.

**Comment by RylanSchaeffer** [comment]
You should clean up the PR then :)

**Comment by badwriter123** [comment]
Yeah I'll do that and get back with the update . 🤗

**Comment by badwriter123** [comment]
Hello @RylanSchaeffer ,
I have removed the unnecessary files from the pr, the required feature implementation can be reviewed.

File: setup.py [hunk 1]
```diff
@@ -102,7 +102,7 @@
     "blobfile",
     "codecarbon>=2.8.1",
     "cookiecutter==1.7.3",
-    "dataclasses",
+    "dataclasses;python_version<'3.7'",
```
Comment (AzizCode92) [inline setup.py]: dataclasses where related to the gemma code I guess here, we should remove it.

**Comment by badwriter123** [comment]
Hello @AzizCode92 
I have removed the unnecessary python version constraint from setup.py and restored it to original state.
Feature Implementation is ready for review.

**Comment by Rocketknight1** [comment]
cc @SunMarc

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -4954,6 +4961,96 @@ def _finish_current_push(self):
             logger.info("Waiting for the current checkpoint push to be finished, this might take a couple of minutes.")
             self.push_in_progress.wait_until_done()
 
+
+    def _compute_mean_token_accuracy(self, logits, labels, model_type='lm', ignore_index=-100, num_virtual_tokens=0):
```
Comment (AzizCode92) [inline src/transformers/trainer.py:4965]: I think we need to seperate between the training and the evaluation mode so the mean_token_accuracy is not mixed up between the two.

Comment (AzizCode92) [reply src/transformers/trainer.py:4965]: I think we need also to add the mean token accuracy to the log method. We need to have here as well seperate containers for the evaluation and the training metrics.
I can think of this pseudo code here
```python 

    def log(self, logs: dict[str, float], start_time: Optional[float] = None) -> None:
        if self.state.epoch is not None:
            logs["epoch"] = self.state.epoch
        # ... (other existing logging logic) ...
  
        if getattr(self.args, "compute_mean_token_accuracy", False):
            mode = "train" if self.model.training else "eval"
            if self._metrics[mode]:
                # Average metrics collected since the last log
                metrics = {key: sum(val) / len(val) for key, val in self._metrics[mode].items()}
                
                # Add "eval_" prefix during evaluation
                if mode == "eval":
                    metrics = {f"eval_{key}": val for key, val in metrics.items()}

                logs.update(metrics)
                # Clear the cache for the next logging interval
                self._metrics[mode].clear()
        
        output = {**logs, **{"step": self.state.global_step}}
        self.state.log_history.append(output)
        self.control = self.callback_handler.on_log(self.args, self.state, self.control, logs)
```

Comment (AzizCode92) [reply src/transformers/trainer.py:4965]: Also thanks for putting me as a reviewer, but this is all I can say here.
The authors of this code can direct you even better.

Comment (badwriter123) [reply src/transformers/trainer.py:4965]: Hello @AzizCode92 ,
The author of this issue mentioned that he is not affiliated with the code and only requires the mean token accuracy feature, so the further enhancements are not needed I guess.
Thanks for your insights on improving the feature. 🤗

**Comment by badwriter123** [comment]
Hello @Rocketknight1 ,
Requesting review of mean token accuracy feature #40303 .

**Comment by qgallouedec** [comment]
Out of curiosity, what are using the trainer for?
In case it's relevant, the SFTTrainer already computes and report the mean token accuracy.

**Comment by RylanSchaeffer** [comment]
Pretraining smallish autoregressive models. To be honest, I don't know what the difference is between transformers trainer and TRL's sft trainer, and I haven't found the time to investigate 😅

**Comment by RylanSchaeffer** [comment]
Pretraining smallish autoregressive models. To be honest, I don't know what the difference is between transformers trainer and TRL's sft trainer, and I haven't found the time to investigate 😅

**Comment by qgallouedec** [comment]
`SFTTrainer` is `Trainer` with some extra features, like preparing your dataset if needed, packing, log useful metrics like the mean token accuracy. Technically you can use `SFTTrainer` to do pretraining.

Let's wait what @SunMarc think of upstreaming the mean token accuracy computation in `Trainer`. I don't have strong opinion on this.

**Comment by SunMarc** [comment]
`SFTTrainer is Trainer with some extra features, like preparing your dataset if needed, packing, log useful metrics like the mean token accuracy. Technically you can use SFTTrainer to do pretraining.`

Can you try to use SFTTrainer to train your model @RylanSchaeffer ? I prefer not adding this if this not very important feature as this will make Trainer more bloated

**Comment by RylanSchaeffer** [comment]
@SunMarc a question and a comment: 

1. Is there any reason to prefer Trainer over SFTTrainer for pretraining?

2. From a user perspective, this is _very_ confusing. As I understand, your guidance is, "For pretraining, install a library called Transformer _Reinforcement Learning_ and then use its SFTTrainer." I don't think using an RL library's SFT implementation for pretraining is intuitive or stated anywhere in any documentation (but if I'm wrong, please let me know). I love TRL and have even made some contributions to it, but I'm a bit baffled by this conceptual organization across libraries.

**Comment by RylanSchaeffer** [comment]
Also, is packing not implemented for Trainer, only SFTTrainer?

**Review by MekkCyber — COMMENTED** [review]
Thanks for your contribution @badwriter123 🤗 ! can you revert the changes deleting some unrelated files like `modeling_gemma3n.py`

**Comment by badwriter123** [comment]
Hello @MekkCyber ,
Files Have been restored as previous.

**Comment by qgallouedec** [comment]
> Is there any reason to prefer Trainer over SFTTrainer for pretraining?

SFTTrainer only add features, so I'd say no. But if you don't use any of these features, then you should use Trainer.

> From a user perspective, this is very confusing. As I understand, your guidance is, "For pretraining, install a library called Transformer Reinforcement Learning and then use its SFTTrainer." 

I'd phrase a bit differently:
For pretraining, I recommend using nanotron. However, if you want to use `transformers.Trainer`, You can sneak around and use `SFTTrainer` directly to benefit from some of its added features (like packing, accuracy logging...), while remaining aware that this is a roundabout way of using it.
